MParT/ParameterizedFunctionBase.h

```diff
@@ -186,6 +186,13 @@ namespace mpart {
         /** Checks to see if the coefficients have been initialized yet, returns true if so, false if not */
         bool CheckCoefficients() const;
 
+        /**
+         * @brief Returns the indices of coeffs for functions non-constant in the last input for their respective output.
+         * 
+         * @return std::vector<unsigned int> 
+         */
+        virtual std::vector<unsigned int> DiagonalCoeffIndices() const;
```

Comment:
I'm not sure it makes sense to expose this in `ParameterizedFunctionBase` because the function could be rectangular, is generally not triangular, and the definition of what is the "diagonal" might not be as well-defined as it is for `MonotoneComponent` or `TriangularMap`.   Is there anywhere we need this in PFB or can we always cast to a child with a well-defined diagonal?

Reply:
In C++ land, you're likely correct. However, in Julia/Python/Matlab land (or at least julia-land), the answer is a resounding "unsure". There's actually a bigger corner case of `ComposedMap`, which is just a disaster in- and of- itself (Suppose you compose a triangular map with a sparse affine map; then the diagonal indices would come from some graph algorithm). I agree with you that this is not a "good" way of exposing this functionality. On the other hand, a) since this doesn't make sense for nontriangular maps, I don't see the actual damage; b) I suspect there are very few working with "rectangular" maps that aren't composed; and c) working with bindings makes any "downcasting" from conditional map base questionably difficult and I'm not enthusiastic about an exercise in now needing to cast maps between different levels in the class hierarchy for different functionality as well as getting that working in three binding languages.

Reply:
I don't really like the idea of adding things to the base class just out of convenience for the bindings, but if it would be a royal pain otherwise than I could be convinced.  At least for python, I don't really see any issues.  pybind11 is very good at handling these things and figuring our what child type the objective is.